hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/ssl/ReloadingX509KeystoreManager.java

```diff
@@ -0,0 +1,151 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.security.ssl;
+
+import org.apache.hadoop.classification.InterfaceAudience;
+import org.apache.hadoop.classification.InterfaceStability;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.net.ssl.*;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.Socket;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.security.GeneralSecurityException;
+import java.security.KeyStore;
+import java.security.Principal;
+import java.security.PrivateKey;
+import java.security.cert.X509Certificate;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * An implementation of <code>X509KeyManager</code> that exposes a method,
+ * {@link #loadFrom(Path)} to reload its configuration. Note that it is necessary
+ * to implement the <code>X509ExtendedKeyManager</code> to properly delegate
+ * the additional methods, otherwise the SSL handshake will fail.
+ */
+@InterfaceAudience.Private
+@InterfaceStability.Evolving
+public class ReloadingX509KeystoreManager extends X509ExtendedKeyManager {
+
+    private static final Logger LOG = LoggerFactory.getLogger(ReloadingX509TrustManager.class);
+
+    static final String RELOAD_ERROR_MESSAGE =
+            "Could not load keystore (keep using existing one) : ";
+
+    private String type;
+    private String storePassword;
+    private String keyPassword;
+    private AtomicReference<X509ExtendedKeyManager> keyManagerRef;
+
+    /**
+     * Construct a <code>Reloading509KeystoreManager</code>
+     *
+     * @param type type of keystore file, typically 'jks'.
+     * @param location local path to the keystore file.
+     * @param storePassword password of the keystore file.
+     * @param keyPassword The password of the key.
+     * @throws IOException
+     * @throws GeneralSecurityException
+     */
+    public ReloadingX509KeystoreManager(String type, String location,
+                                        String storePassword, String keyPassword)
+            throws IOException, GeneralSecurityException {
+        this.type = type;
+        this.storePassword = storePassword;
+        this.keyPassword = keyPassword;
+        keyManagerRef = new AtomicReference<X509ExtendedKeyManager>();
+        keyManagerRef.set(loadKeyManager(Paths.get(location)));
+    }
```

Comment:
Yeah, tab seems to be 4 when should be 2 spaces.

Reply:
Fixed. Also fixed line length issues in this file.